Use MiniWallet in functional test rpc_signrawtransactionwithkey.

martinsaposnic opened the pull request:
In response to issue https://github.com/bitcoin/bitcoin/issues/30600, optimizations have been implemented to enhance test efficiency and readability:

This PR refactors the `rpc_signrawtransactionwithkey.py` functional test to use MiniWallet for creating funding transactions. This simplifies the test code and improves performance by eliminating the need to mine new blocks for each funding transaction.

Key changes:
- Replaced custom `send_to_address` method with MiniWallet's `send_to` method
- Removed unnecessary setup of a clean chain and second node
- Simplified transaction creation and signing process

theStack reviewed (commented):
Thanks for picking this up!

A few more suggestions to reduce the code even further:
* the initialization of previously needed instance members (`self.blk_idx`, `self.block_hash`) can now be removed
* instead of starting with a clean chain (see `self.setup_clean_chain`), you could use the pre-generated test chain to avoid the need of generating blocks; MiniWallet will rescan and pick up the relevant UTXOs up automically at instantiation. IIUC it shouldn't be necessary to mine any blocks for this tests
* regarding the P2A test change: MiniWallet's `send_to` returns both the txid and the output index, so there is no need to find it manually after; you could e.g. return a tuple (txid,vout) from `send_to_address`

martinsaposnic commented:
Out of curiosity I ran a little benchmark 

Master:

<img width="607" alt="image" src="https://github.com/user-attachments/assets/875f96fd-5ef2-4730-a3df-f7b3ff40f039">


With MiniWallet:

<img width="903" alt="image" src="https://github.com/user-attachments/assets/40bdcb95-1b08-4054-9223-44566bdca548">

Miniwallet is cool

martinsaposnic commented:
@theStack I implemented the 3 suggestions. Let me know if you have further comments 🙂

theStack reviewed (commented):
lgtm, once again some smaller suggestions

martinsaposnic commented:
@theStack thank you for the suggestions! just forcePushed a new commit with the changes.

theStack reviewed (approved):
lgtm ACK 96d34fb2e493a2c37907e76808215c5ab0c3cf70

ismaelsadeeq reviewed (commented):
Concept ACK

I'll prefer if you split this commit intro three as you highlight in commit message

1- Replace custom funding tx creation with MiniWallet
2- Remove unnecessary clean chain setup and second node
2- Simplify transaction handling

martinsaposnic committed:
Replace custom funding tx creation with MiniWallet.

setup_clean_chain=True is deleted so it uses the default.
Also, vout is now returned from send_to_address,
so now there is no need to fetch it manually

Also remove not-needed code that was used with the old
transaction handling.

martinsaposnic committed:
Remove second node since only 1 is needed for the test

martinsaposnic commented:
@ismaelsadeeq rebased and pushed. thanks for the review!

ismaelsadeeq commented on test/functional/rpc_signrawtransactionwithkey.py:
```diff
@@ -46,16 +49,12 @@
-        txid = self.nodes[0].sendrawtransaction(self.nodes[0].signrawtransactionwithkey(rawtx, [self.nodes[0].get_deterministic_priv_key().key])["hex"], 0)
-        return txid
+        script_pub_key = address_to_scriptpubkey(addr)
+        tx = self.wallet.send_to(from_node=self.nodes[0],scriptPubKey=script_pub_key, amount=int(amount * COIN))
```
nit: add spaces between passed arguments
 ```suggestion
         tx = self.wallet.send_to(from_node=self.nodes[0], scriptPubKey=script_pub_key, amount=int(amount * COIN))
```

ismaelsadeeq replied:
Why are you converting the amount to an integer here? Won’t that change the assumptions in the test? I see 49.999 explicitly passed in the test—don’t we want to maintain precision?

martinsaposnic replied:
I noticed that MiniWallet's `send_to` method doesn't play nice with non-integer amounts. It's failing because CTxOut's `nValue` is an int, so when you pass a float, the `to_bytes` method throws an error.

I did a quick scan of the codebase:

We've got 21 uses of MiniWallet's `send_to` method on the tests (not counting this PR)
- 17 of these are already sending an integer
- 4 are explicitly doing int(amount) before passing it to `send_to`

I think fixing this is probably outside what we're trying to do in this PR. I'm happy to create a separate issue for it and tackle it. 

@ismaelsadeeq, what do you think?

martinsaposnic replied:
@theStack what do you think?

ismaelsadeeq replied:
> It's failing because `CTxOut`'s `nValue` is an `int`, so when you pass a `float`, the `to_bytes` method throws an error.

I can also verify that this fails locally.

I'm generally not a fan of magic numbers, but the obvious explanation is that some satoshis were subtracted from the UTXO value to account for transaction fees.

If that's the case, then it's fine to round it down.

> I'm happy to create a separate issue for it and tackle it.

+1

ismaelsadeeq commented on test/functional/rpc_signrawtransactionwithkey.py:
```diff
@@ -46,16 +49,12 @@
-        return txid
+        script_pub_key = address_to_scriptpubkey(addr)
+        tx = self.wallet.send_to(from_node=self.nodes[0],scriptPubKey=script_pub_key, amount=int(amount * COIN))
+        return [tx["txid"], tx["sent_vout"]]
```
It might be better to return a tuple here
```suggestion
        return tx["txid"], tx["sent_vout"]
```

ismaelsadeeq commented on test/functional/rpc_signrawtransactionwithkey.py:
```diff
@@ -4,9 +4,6 @@
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
```
8c64048c443cddce201b6d54184b32f3894ab806 can be squashed together with the first commit

ismaelsadeeq reviewed (commented):
code review ACK a563f41232e2dd360ee8e76f6348dd10c7f4f2a3

theStack reviewed (approved):
ACK a563f41232e2dd360ee8e76f6348dd10c7f4f2a3

glozow commented:
ACK a563f41232e2dd360ee8e76f6348dd10c7f4f2a3
